### PR TITLE
New Script: SilverBullet

### PR DESCRIPTION
**Pull request by dsiebel**
## Description

I was looking into Silverbullet today and while installing it in a debian LXC I created this install and update script. Since Silverbullet was already requested (#127) and it's already on the backlog, I figured: why not.

Fixes #127 

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [x] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [x] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [x] Testing performed (I have tested my changes, ensuring everything works as expected)
- [x] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)

I intentionally decided against using the `silverbullet upgrade` command, since this would require installing and running the entire `deno` build. Using pre-built binary just feels better for production use.

Ref: https://silverbullet.md/Install/Deno#Upgrading

## Related Pull Requests / Discussions

If there are other pull requests or discussions related to this change, please link them here:
- Related PR #

### CODE REVIEW

**Comment by MickLesk** [comment]
It's not worth checking every line for now, there are so many errors and things that don't fit the conventions of our system.

Please check correspondence objects such as kimai, snipe-it, komga, vikunja, bookstack etc.

**Comment by dsiebel** [comment]
> It's not worth checking every line for now, there are so many errors and things that don't fit the conventions of our system.
> 
> Please check correspondence objects such as kimai, snipe-it, komga, vikunja, bookstack etc.

I could really use some pointers then. I used the scripts to install and update Silverbullet on my running Debian LXC, so I am wondering what errors there are.

To be quite frank, it is impossible at the moment to find "the right way" to do things, and I get the feeling it keeps changing with every PR I create..
I looked at `traefik`, `trilium`, `debian` and `zoraxy`, which I recently updated, when creating the scripts.

I _did_ add some improvements (or so I thought), like checking the service is actually running after install and update. Is this what you are referring to?

**Comment by dsiebel** [comment]
To visualize my point, this is a side-by-side comparison of the updated scripts between `vikunja` and the newly added `silverbullet`. The boilerplate code is all the same, the things that are different are because it's a different app that's being installed.

image

File: ct/silverbullet.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  SSH="no"
+  VERB="no"
+  echo_default
+}
+
+function update_script() {
+  header_info
+  check_container_storage
+  check_container_resources
+
```
Comment (MickLesk) [inline ct/silverbullet.sh:59]: Remove blank line

File: ct/silverbullet.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  service_path="/etc/systemd/system/silverbullet.service"
```
Comment (MickLesk) [inline ct/silverbullet.sh:60]: Nope, We don't use anything like that.

File: ct/silverbullet.sh [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  if [[ ! -d /opt/silverbullet || ! -f "${service_path}" ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
+
```
Comment (MickLesk) [inline ct/silverbullet.sh:62]: Remove blank line

File: ct/silverbullet.sh [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  echo_default
+}
+
+function update_script() {
+  header_info
+  check_container_storage
+  check_container_resources
+
+  service_path="/etc/systemd/system/silverbullet.service"
+  if [[ ! -d /opt/silverbullet || ! -f "${service_path}" ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
```
Comment (MickLesk) [inline ct/silverbullet.sh:61]: Do it the same way as in other scripts

File: ct/silverbullet.sh [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  RELEASE=$(curl -s https://api.github.com/repos/silverbulletmd/silverbullet/releases/latest | grep "tag_name" | awk '{print substr($2, 2, length($2)-3) }')
+  if [[ ! -f "/opt/silverbullet/${APP}_version.txt" || "${RELEASE}" != "$(cat /opt/silverbullet/${APP}_version.txt)" ]]; then
```
Comment (MickLesk) [inline ct/silverbullet.sh:64]: Wrong folder

File: ct/silverbullet.sh [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    msg_info "Stopping ${APP}"
+    systemctl stop silverbullet.service
```
Comment (MickLesk) [inline ct/silverbullet.sh:66]: .service Not needed

File: ct/silverbullet.sh [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    msg_ok "Stopped ${APP}"
+
+    msg_info "Updating ${APP} to ${RELEASE}"
```
Comment (MickLesk) [inline ct/silverbullet.sh:69]: v${RELEASE}

File: ct/silverbullet.sh [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    wget -q https://github.com/silverbulletmd/silverbullet/releases/download/${RELEASE}/silverbullet-server-linux-x86_64.zip
+    unzip silverbullet-server-linux-x86_64.zip &>/dev/null
+    mv silverbullet /opt/silverbullet/bin/
+    chmod +x /opt/silverbullet/bin/silverbullet
+    ln -sf /opt/silverbullet/bin/silverbullet /usr/local/bin/silverbullet
```
Comment (MickLesk) [inline ct/silverbullet.sh:74]: For what?

File: ct/silverbullet.sh [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    echo "${RELEASE}" >/opt/silverbullet/${APP}_version.txt
+    msg_ok "Updated ${APP}"
```
Comment (MickLesk) [inline ct/silverbullet.sh:76]: Remove

File: ct/silverbullet.sh [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+    msg_info "Starting ${APP}"
+    systemctl start silverbullet.service
```
Comment (MickLesk) [inline ct/silverbullet.sh:79]: .service not needed

File: ct/silverbullet.sh [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    sleep 1
```
Comment (MickLesk) [inline ct/silverbullet.sh:80]: Not needed

File: ct/silverbullet.sh [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    if systemctl status silverbullet.service &>/dev/null ; then
+      msg_ok "Started ${APP}"
+      msg_ok "Updated Successfully"
```
Comment (MickLesk) [inline ct/silverbullet.sh:83]: updated app to v${RELEASE}

File: ct/silverbullet.sh [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    else
+      msg_error "Failed to start ${APP}"
+    fi
```
Comment (MickLesk) [inline ct/silverbullet.sh:86]: Not needed

File: install/silverbullet-install.sh [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+source /dev/stdin <<< "$FUNCTIONS_FILE_PATH"
+color
+verb_ip6
+catch_errors
+setting_up_container
+network_check
+update_os
+
+msg_info "Installing Dependencies"
+$STD apt-get install -y curl
```
Comment (MickLesk) [inline install/silverbullet-install.sh:17]: Harmonize the whole Block to other current Scripts

File: install/silverbullet-install.sh [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+$STD apt-get install -y sudo
+$STD apt-get install -y mc
+msg_ok "Installed Dependencies"
+
+RELEASE=$(curl -s https://api.github.com/repos/silverbulletmd/silverbullet/releases/latest | grep "tag_name" | awk '{print substr($2, 2, length($2)-3) }')
+
+msg_info "Installing ${APPLICATION}"
```
Comment (MickLesk) [inline install/silverbullet-install.sh:24]: App

File: install/silverbullet-install.sh [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+mkdir -p /opt/silverbullet/bin /opt/silverbullet/space
+wget -q https://github.com/silverbulletmd/silverbullet/releases/download/${RELEASE}/silverbullet-server-linux-x86_64.zip
+unzip silverbullet-server-linux-x86_64.zip &>/dev/null
```
Comment (MickLesk) [inline install/silverbullet-install.sh:27]: -q WITHOUT &>/dev/null & you can unzip it in the correct folder without the next line

File: install/silverbullet-install.sh [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+mv silverbullet /opt/silverbullet/bin/
+chmod +x /opt/silverbullet/bin/silverbullet
+ln -s /opt/silverbullet/bin/silverbullet /usr/local/bin/silverbullet
+echo "${RELEASE}" >/opt/silverbullet/${APPLICATION}_version.txt
```
Comment (MickLesk) [inline install/silverbullet-install.sh:31]: Only in /opt

File: install/silverbullet-install.sh [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+RELEASE=$(curl -s https://api.github.com/repos/silverbulletmd/silverbullet/releases/latest | grep "tag_name" | awk '{print substr($2, 2, length($2)-3) }')
+
+msg_info "Installing ${APPLICATION}"
+mkdir -p /opt/silverbullet/bin /opt/silverbullet/space
+wget -q https://github.com/silverbulletmd/silverbullet/releases/download/${RELEASE}/silverbullet-server-linux-x86_64.zip
+unzip silverbullet-server-linux-x86_64.zip &>/dev/null
+mv silverbullet /opt/silverbullet/bin/
+chmod +x /opt/silverbullet/bin/silverbullet
+ln -s /opt/silverbullet/bin/silverbullet /usr/local/bin/silverbullet
```
Comment (MickLesk) [inline install/silverbullet-install.sh:30]: For what?

File: install/silverbullet-install.sh [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+echo "${RELEASE}" >/opt/silverbullet/${APPLICATION}_version.txt
+msg_ok "Installed ${APPLICATION}"
```
Comment (MickLesk) [inline install/silverbullet-install.sh:32]: APP

File: install/silverbullet-install.sh [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+msg_info "Creating Service"
+service_path="/etc/systemd/system/silverbullet.service"
```
Comment (MickLesk) [inline install/silverbullet-install.sh:35]: Nope

File: install/silverbullet-install.sh [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+echo "[Unit]
```
Comment (MickLesk) [inline install/silverbullet-install.sh:37]: Cat EOF not echo

File: install/silverbullet-install.sh [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+systemctl enable --now -q silverbullet
+msg_ok "Created Service"
+
+msg_info "Starting ${APPLICATION}"
+systemctl start silverbullet.service
+sleep 1
+if systemctl status silverbullet.service &>/dev/null ; then
+	msg_ok "Started ${APPLICATION}"
+else
+	msg_error "Failed to start ${APPLICATION}"
```
Comment (MickLesk) [inline install/silverbullet-install.sh:59]: Not needed, enable -q --now already started the app

File: json/silverbullet.json [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+    "slug": "silverbullet",
+    "categories": [
+        14
+    ],
+    "date_created": "2024-12-03",
+    "type": "ct",
+    "updateable": true,
+    "privileged": false,
+    "interface_port": 3000,
+    "documentation": null,
```
Comment (MickLesk) [inline json/silverbullet.json:12]: No docu?

**Comment by MickLesk** [comment]
Short review only with the mobile phone. We don't do the whole thing with services etc. like this. In general, take a look at current scripts and not the ones you mentioned above, which have not been revised for ages.

**Comment by dsiebel** [comment]
Thanks for the review! Will address asap.

File: ct/silverbullet.sh (hunk 8 above)
Comment (dsiebel) [reply ct/silverbullet.sh:74]: Mostly convenience.
Running `silverbullet --help` vs `opt/silverbullet/bin/silverbullet --help`

File: install/silverbullet-install.sh (hunk 15 above)
Comment (dsiebel) [reply install/silverbullet-install.sh:24]: `$APP` is only available in update script, if I am not mistaken.
In install scripts it needs to be `$APPLICATION`.

File: install/silverbullet-install.sh (hunk 19 above)
Comment (dsiebel) [reply install/silverbullet-install.sh:32]: See above, didn't work for me, only `$APPLICATION` was available in install scripts.

File: install/silverbullet-install.sh (hunk 15 above)
Comment (dsiebel) [reply install/silverbullet-install.sh:25]: Double checked with the scripts you mentioned, to be super-sure:

- kimai: `APPLICATION`
- komga: `APPLICATION`
- snipeit: none, hardcoded `snipe-it` or `snipeit`
- vikunia: `APPLICATION`
- bookstack: both `APPLICATION` and hard-coded `bookstack`

File: json/silverbullet.json (hunk 23 above)
Comment (dsiebel) [reply json/silverbullet.json:12]: I checked a bunch of other JSONs and all had `documentation: null`, thought the field might be unused.
Guess I was unlucky.
Either way, the website is the documentation so I linked a sub-page, that seemed fitting.

File: ct/silverbullet.sh [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+  MTU=""
+  SD=""
+  NS=""
+  MAC=""
+  VLAN=""
+  SSH="no"
+  VERB="no"
+  echo_default
+}
+
```
Comment (MickLesk) [inline ct/silverbullet.sh:54]: can you migrate to new header? Example:

```bash
#!/usr/bin/env bash
source <(curl -s https://raw.githubusercontent.com/community-scripts/ProxmoxVE/main/misc/build.func)
# Copyright (c) 2021-2024 tteck
# Author: MickLesk (Canbiz) & vhsdream
# License: MIT | https://github.com/community-scripts/ProxmoxVE/raw/main/LICENSE
# Source: https://hoarder.app/

# App Default Values
APP="Hoarder"
var_tags="bookmark"
var_cpu="2"
var_ram="4096"
var_disk="8"
var_os="debian"
var_version="12"
var_unprivileged="1"

# App Output & Base Settings
header_info "$APP"
base_settings

# Core
variables
color
catch_errors
```

File: ct/silverbullet.sh [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+  exit
+}
+
+start
+build_container
+description
+
+msg_ok "Completed Successfully!\n"
+echo -e "${APP} should be reachable by going to the following URL.
+         ${BL}http://${IP}:3000${CL} \n"
```
Comment (MickLesk) [inline ct/silverbullet.sh:89]: can you migrate to new footer? Example:
```bash
msg_ok "Completed Successfully!\n"
echo -e "${CREATING}${GN}${APP} setup has been successfully initialized!${CL}"
echo -e "${INFO}${YW} Access it using the following URL:${CL}"
echo -e "${TAB}${GATEWAY}${BGN}http://${IP}:3000${CL}"
```

File: ct/silverbullet.sh (hunk 24 above)
Comment (dsiebel) [reply ct/silverbullet.sh]: done.

**Comment by dsiebel** [comment]
I just noticed, the project card for this wasn't moved to "In PR" yet. Can someone take care of it for transparency?

**Review by michelroegl-brunner — APPROVED** [review]
LGTM

**Review by MickLesk — APPROVED** [review]

**Pull request merged**